### PR TITLE
8336934: Clean up JavaLangReflectAccess

### PR DESCRIPTION
**Pull request by liach**
Removed redundant APIs in `JavaLangReflectAccess` and added general warning against using `SharedSecrets`.

The cleanup in `JavaLangReflectAccess` is:
- Converted `newConstructor` taking parameters to taking another constructor and a new accessor. The old existing API essentially does a copy and sets an accessor.
- Removed lots of constructor getters, unused after `newConstructor` cleanup.
- Removed accessor getter/setters, unused after `newConstructor` cleanup.
- Removed `leafCopyMethod`. We can already use `getRoot` plus `copyMethod` to achieve the same effect.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8336934](https://bugs.openjdk.org/browse/JDK-8336934): Clean up JavaLangReflectAccess (**Enhancement** - P4)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20290/head:pull/20290` \
`$ git checkout pull/20290`

Update a local copy of the PR: \
`$ git checkout pull/20290` \
`$ git pull https://git.openjdk.org/jdk.git pull/20290/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20290`

View PR using the GUI difftool: \
`$ git pr show -t 20290`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20290.diff">https://git.openjdk.org/jdk/pull/20290.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20290#issuecomment-2244218255)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/access/JavaLangReflectAccess.java [hunk 1]
```diff
@@ -29,73 +29,43 @@
...
     /** Gets the shared array of parameter types of an Executable. */
     public Class<?>[] getExecutableSharedParameterTypes(Executable ex);
 
     /** Gets the shared array of exception types of an Executable. */
     public Class<?>[] getExecutableSharedExceptionTypes(Executable ex);
 
-    //
+    //</editor-fold>
+
+    //<editor-fold defaultstate="collapsed" desc="AccessibleObject Copying">
```
Comment (RogerRiggs) [inline src/java.base/share/classes/jdk/internal/access/JavaLangReflectAccess.java:53]: Personal editor comments should not be included.

File: src/java.base/share/classes/jdk/internal/access/JavaLangReflectAccess.java [hunk 2]
```diff
@@ -29,73 +29,43 @@
...
-    public Method      leafCopyMethod(Method arg);
-
     /** Makes a "child" copy of a Field */
     public Field       copyField(Field arg);
 
     /** Makes a "child" copy of a Constructor */
     public <T> Constructor<T> copyConstructor(Constructor<T> arg);
 
+    //</editor-fold>
+
```
Comment (RogerRiggs) [inline src/java.base/share/classes/jdk/internal/access/JavaLangReflectAccess.java:68]: Remove

**Comment by RogerRiggs** [comment]
The description should include a summary of the changes and the rationale.
The Jira description is a bit better and could also be more complete.

File: src/java.base/share/classes/java/lang/reflect/Constructor.java [hunk 3]
```diff
@@ -165,6 +165,14 @@ Constructor<T> copy() {
         return res;
     }
 
+    /** Creates a new root constructor with a custom accessor for serialization hooks. */
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/reflect/Constructor.java:168]: Preserve "/**" comments for generated javadoc, a "//" comment would be more appropriate for a 1-liner.

File: src/java.base/share/classes/jdk/internal/access/JavaLangReflectAccess.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/jdk/internal/access/JavaLangReflectAccess.java:53]: Ah, thought editor folds were acceptable as they appear in a few places in javac code. Will remove.

**Comment by liach** [comment]
@RogerRiggs Is the description now cleaner?

**Comment by RogerRiggs** [comment]
/reviewers 2 reviewer
Another person familiar with the reflection implementation should review this also

**Review by RogerRiggs — APPROVED** [review]
looks good to me.
Thanks for the comments and updated description.

**Review by jddarcy — APPROVED** [review]
Looks fine.

**Comment by liach** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**